mlflow/pyfunc/__init__.py

```diff
@@ -941,7 +1005,91 @@ def predict(*args):
         else:
             return result[result.columns[0]]
 
-    udf = pandas_udf(predict, result_type)
+    def predict(iterator):
+        # importing here to prevent circular import
+        from mlflow.pyfunc import scoring_server
+        from mlflow.pyfunc.scoring_server.client import ScoringServerClient, prepare_env
+
+        # Note: this is a pandas udf function in iteration style, which takes an iterator of
+        # tuple of pandas.Series and outputs an iterator of pandas.Series.
+
+        scoring_server_proc = None
+
+        # TODO: Support virtual env.
+        #
+        # TODO: For conda/virtualenv restored env cases,
+        #  For each individual python process (driver side), create individual and temporary
+        #  conda env dir / virtualenv env dir and when process exit,
+        #  delete the temporary env dir.
+        #  The reason is
+        #   1. env dir might be a large size directory and cleaning it when process exit
+        #      help saving disk space.
+        #   2. We have conda package cache dir and pip cache dir which are shared across all
+        #      python processes which help reducing downloading time.
+        #   3. Avoid race conditions related issues.
+        #
+        # TODO:
+        #   For NFS available case, set conda env dir / virtualenv env dir in sub-directory under
+        #   NFS directory, and in spark driver side prepare restored env once, and then all
+        #   spark UDF tasks running on spark workers can skip re-creating the restored env.
+        if env_manager == "conda":
```

Comment:
Nit: Can we break these into separate functions?

Reply:
Shall we need to split them ? local / conda mode share many parts of code.

Reply:
We can refactor it in better way, I think we can consider doing that after we add vitualenv support ?

Reply:
Sounds good!